storage.go

```diff
@@ -100,11 +102,6 @@ type Locker interface {
 	UnlockAllObtained()
```

Comment:
This doesn't appear to be called from anywhere and I don't think it's needed anyway so can be removed.

Reply:
This is called by Caddy during shutdown (any signal is not a forceful shutdown).

Reply:
This method doesn't work for cleanup at program start, so it's used only during shutdown and giving this some more thought, I think it can lead to double-unlocking and other bugs:

* Unless it's an immediate `os.Exit()` (an even then), I don't think it's safe to unlock any of the keys: If we unlock the key, the code that locked it will continue running and will eventually call unlock again.
* The existing running code might also go on to change some state, but another process can now acquire the lock and proceed to change the same state.
* Given the above points, it's unsafe to unlock keys locked by our own process (outside the code that locked it).
* The only case (I can think of) in which our process's code could fail to unlock the key is if it panics, and we have `defer` for that.
* The simplest solution I can think of would be to just wait for all code to come back to a safe point (after unlock) - maybe with some context telling it we're shutting down. At that point there shouldn't be any case in which `UnlockAllObtained()` has anything to unlock.

After reading the code some more I saw a couple more bugs. I will comment on those separately.

Reply:
Interesting thoughts -- I'd be happy to review a potential solution if you want to submit one!

Reply:
I eliminated this method.